text/3556-re-export-symbols.md

```diff
@@ -0,0 +1,175 @@
+
+When linking a staticlib into a cdylib crate, there are some cases where it is useful to re-export symbols. There is extensive dicussion from [people wanting to do this](https://github.com/rust-lang/rfcs/issues/2771) (with a note that the Lang team discussed this in 2017!), as well as another issue from people attempting to [use `+whole-archive`](https://github.com/rust-lang/rust/issues/110624) to re-export _all_ symbols from a linked staticlib.
+
+However, the *only* supported way to re-export a symbol today is manually write wrappers around the staticlib symbols, re-passing the arguments down. This is:
+ - inconvenient - re-exporting as the same name requires putting the `extern` block and wrapper function in different modules due to name collision
+ - error-prone - the user has to maintain the wrapper function with the correct arguments
+ - sometimes impossible - variadic functions are [not stable yet](https://github.com/rust-lang/rust/issues/44930)
+
+This tightly-scoped RFC adds support for re-exporting explicitly listed symbols from a linked staticlib in an extern block to the final produced cdylib. It does *not* add 'wildcard' support to support re-exporting all symbols from a staticlib, nor does it cover other staticlib/cdylib combinations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When producing a cdylib from a Rust crate and linking to a staticlib, adding `#[no_mangle]` and `pub` to items in an extern block in the top level crate as follows:
+
+```
+// kind is optional if it's been specified elsewhere, e.g. via the `-l` flag to rustc
+#[link(name="ext", kind="static")]
+extern {
+    #[no_mangle]
+    pub fn foo();
+
+    #[no_mangle]
+    pub static bar: std::ffi::c_int;
+}
+```
+
+will make the `foo` and `bar` symbols from the static library (e.g. `libext.a` on Linux) visible in the final produced `libcrate.so`. This is supported for all item types supported in an `extern` block - at time of writing, this means `fn`s and `static`s.
+
+If the crate is not being build as a cdylib the `#[no_mangle]` annotation will be ignored - this includes when the crate is a dependency of a cdylib crate.
+
+If `ext` is not _explicitly_ linked as a static library the result is unspecified - the symbol may or may not be visible depending on the behavior of the linker. A non-fatal warning will be emitted:
+
+```
+warning: `#[no_mangle]` with `pub` is only valid for re-export when the library is explicitly a `staticlib`
+```
+
+If the item is marked as `#[no_mangle]` without being publicly visible in the crate (e.g. the `pub` annotation is missing or the `extern` block is in a private module), the `#[no_mangle]` annotation will have no effect. A non-fatal warning will be emitted:
+
+```
+warning: `#[no_mangle]` will not re-export private items from `extern` block
+```
+
+# Reference-level explanation
```

Comment:
In light of https://github.com/rust-lang/rfcs/pull/3325, have you considered whether the attribute should be `unsafe` or not?

Reply:
The motivating issue there suggests that allowing rust crates to export symbols at all is unsafe, so (assuming that RFC gets merged) whatever attribute this RFC ends up using would presumably also be unsafe.

If that RFC gets merged first then I'd update this one as needed, or if this goes first then I assume there will be a pass over attributes to see if they should be unsafe.